clang/lib/Parse/ParseDecl.cpp

```diff
@@ -5676,24 +5676,32 @@ Parser::DeclGroupPtrTy Parser::ParseTopLevelStmtDecl() {
   // Parse a top-level-stmt.
   Parser::StmtVector Stmts;
   ParsedStmtContext SubStmtCtx = ParsedStmtContext();
-  Actions.PushFunctionScope();
+  ParseScope FnScope(this, Scope::FnScope | Scope::DeclScope |
```

Comment:
I was wondering if we could survive without this scope here or at least if we could drop the `Scope::CompoundStmtScope` part of it.

Reply:
I tested without it the `ParseScope` and got a segfault, so I assume we can not drop it (without looking into the details). I kept both, `FnScope` and `CompoundStmtScope`, because in `ActOnStartTopLevelStmtDecl()` we push them on the Sema side as well.

Reply:
It's a bit awkward to push scoping state in two places. I am wondering if moving the `FnScope` down in `ActOnStartTopLevelStmtDecl` and `ActOnFinishTopLevelStmtDecl` would be a good idea. It would make the implementation look cleaner but if an error occurs the state of the scopes will be inconsistent (the `R.isUsable` branch). Maybe it is fine as it is...